### PR TITLE
pix_range and pix_db_range in gen_sqw disambuguation

### PR DESCRIPTION
**Pull request by abuts**
This PR is part of ticket on issues with symmetrization, related to correct ranges in gen_sqw. Clear distinction has been drown in gen_sqw where range means what, so variables currying out different meaning have different names. 

At least two small but nasty bugs were fixed. Number of tests (or rather test blocks)  are added  to verify that variables behave correctly

### CODE REVIEW

File: horace_core/utilities/@sqw/symmetrise_sqw.m [hunk 1]
```diff
@@ -146,6 +149,10 @@
 idx = find(side_dot > 0);
 coords_new([1:3], idx) = Reflec*coords_new([1:3], idx); % MP: (TODO) could potentially be optimized further
 clear 'side_dot'; % MP: not needed anymore
+% Testing the ranges transformation
```
Comment (nickbattam-tessella) [inline horace_core/utilities/@sqw/symmetrise_sqw.m:152]: new, commented out code

File: horace_core/utilities/@sqw/symmetrise_sqw.m [hunk 2]
```diff
@@ -122,7 +124,8 @@
 
 %Note that we allow the inclusion of an offset from the origin of the
 %reflection plane. This is specified in rlu.
-vec3=uconv\(v3'-header.uoffset(1:3));
+%vec3=uconv\(v3'-header.uoffset(1:3));
```
Comment (nickbattam-tessella) [inline horace_core/utilities/@sqw/symmetrise_sqw.m:127]: is the header offset no-longer used in this calculation and the vec1 & vec2 calculations above?

File: horace_core/sqw/file_io/class_helpers/@pix_combine_info/pix_combine_info.m [hunk 3]
```diff
@@ -222,8 +246,21 @@
...
             end
         end
+        function pix_range = recalc_pix_range_from_loaders(ldrs)
+            % Recalculate pixels range using list of defined loaders
+            n_files = numel(ldrs);
+            ldr = ldrs{1};
+            pix_range= ldr.get_pix_range();
+            for i=2:n_files
+                ldr = ldrs{i};
+                loc_range = ldr.get_pix_range();
```
Comment (nickbattam-tessella) [inline horace_core/sqw/file_io/class_helpers/@pix_combine_info/pix_combine_info.m:256]: What is `loc_range`?

File: _test/test_gen_sqw_workflow/test_gen_sqw_accumulate_sqw_herbert.m [hunk 4]
```diff
@@ -154,17 +152,27 @@ function test_worker(obj)
             %write_nsqw_to_sqw(infiles,'test_sqw_file.sqw');
             %[main_header,header,datahdr,pos_npixstart,pos_pixstart,npixtot,det,ldrs] = ...
             [~,~,~,~,~,~,det,ldrs] = accumulate_headers_job.read_input_headers(tmp_files);
+            %
+            pix_range = PixelData.EMPTY_RANGE_;
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_gen_sqw_workflow/test_gen_sqw_accumulate_sqw_herbert.m:156]: This tests parallel job on a single thread as if it is running on a remote system. Concurrency is not an issue but all communication protocol is verified. I think it is very useful thing to do, as any logical error in protocol should be exposed here, (except race conditions) and I do not see a way of testing it simpler, as it is the protocol of communicating with a remote system. I have modified the name of the test to  bring a bit more clarity and added comment on what is doing, but it does not make things simpler.

File: _test/test_gen_sqw_workflow/test_gen_sqw_accumulate_sqw_parpool.m [hunk 5]
```diff
@@ -70,8 +70,11 @@ function test_accumulate_and_combine1to4(obj,varargin)
             test_accumulate_and_combine1to4@gen_sqw_accumulate_sqw_tests_common(obj,varargin{:});
         end
         function test_accumulate_sqw1456(obj,varargin) 
-            skipTest('Test test_accumulate_sqw1456 is disabled');
-            % test_accumulate_sqw1456@gen_sqw_accumulate_sqw_tests_common(obj,varargin{:});
+            if ispc
+                skipTest('Test test_accumulate_sqw1456 is disabled, issue #380');
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_gen_sqw_workflow/test_gen_sqw_accumulate_sqw_parpool.m:74]: as the message is printed on PC only, its a bit of duplication, but no harm in it.

File: _test/test_mex_nomex/test_main_mex.m [hunk 6]
```diff
@@ -67,9 +67,9 @@ function tearDown(this)
         
         
         function this=test_accum_cut(this)
-            mex_present=fileparts(which('accumulate_cut_c'));
-            if isempty(mex_present)
-                skipTest('Mex file accumulate_cut_c is not availible on this computer');
+            [~,n_errors] = check_horace_mex();
+            if n_errors>0
+                skipTest('Can not accumulate_cut using mex files');
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_mex_nomex/test_main_mex.m:72]: made skip messages more consistent

File: _test/test_mex_nomex/test_main_mex.m [hunk 7]
```diff
@@ -222,9 +242,10 @@ function test_sort_pix(this)
             pix2 = sort_pix(pix,ix,npix,'-nomex');
             assertElementsAlmostEqual(pix1.data,pix2.data);
             
-            if ~get(hor_config,'use_mex')
-                skiptest('MEX is disabled')
-            end
+            [~,n_errors] = check_horace_mex();            
+            if n_errors>0
+                skipTest('MEX code is broken and can not be used');
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_mex_nomex/test_main_mex.m:247]: the matlab code is tested, only mex code can not. I have modified the message to carry more clarity

File: _test/test_gen_sqw_workflow/test_gen_sqw_accumulate_sqw_herbert.m (hunk 4 above)
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_gen_sqw_workflow/test_gen_sqw_accumulate_sqw_herbert.m:156]: This is input parameter of the procedure, described as an input parameter

File: _test/test_mex_nomex/test_main_mex.m [hunk 8]
```diff
@@ -127,14 +138,19 @@ function tearDown(this)
...
         
         function test_calc_proj_options(this)
-            hcf=hor_config;
-            if ~hcf.use_mex
-                skipTest('MEX is disabled');
-            end
-            cleanup_obj=onCleanup(@()set(hcf,'use_mex',1));
+            [~,n_errors] = check_horace_mex();            
+            if n_errors>0
+                skipTest('MEX code is broken and can not be used');
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_mex_nomex/test_main_mex.m:143]: made skip messages more consistent

File: _test/test_mex_nomex/test_main_mex.m [hunk 9]
```diff
@@ -177,10 +192,12 @@ function test_recompute_bin_data(~)
             assertElementsAlmostEqual(4*s,npix);
             assertElementsAlmostEqual((4*4)*e,npix);
             
+
+            [~,n_errors] = check_horace_mex();            
+            if n_errors>0
+                skipTest('MEX code is broken and can not be used');
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_mex_nomex/test_main_mex.m:198]: made skip messages more consistent and as before

File: horace_core/sqw/@rundatah/private/calc_projections_.m [hunk 10]
```diff
@@ -76,15 +76,17 @@
...
+            
             data = struct('S',obj.S,'ERR',obj.ERR,'en',obj.en,'run_id',obj.run_id);
             det  = obj.det_par;
             efix  = obj.efix;
             emode = obj.emode;
             %proj_mode = 2;
             %nThreads = 1;
             [pix_range,pix] =calc_projections_c(spec_to_cc, data, det, efix,k_to_e, emode, nThreads,proj_mode);
-            pix = PixelData(pix);
+            if proj_mode==2
```
[the remark this answers is not in the document]
Comment (abuts) [reply horace_core/sqw/@rundatah/private/calc_projections_.m:87]: this is the input option of the procedure, described at input

File: horace_core/utilities/@sqw/symmetrise_sqw.m (hunk 1 above)
Comment (abuts) [reply horace_core/utilities/@sqw/symmetrise_sqw.m:152]: this is the part of the work on the ticket with has been left outside the PR. The ticket itself is about problems with symmetrisation, and here I've realized that other work has to be done before going any further. Left here as starting point for further exploration.

File: horace_core/utilities/@sqw/symmetrise_sqw.m (hunk 2 above)
Comment (abuts) [reply horace_core/utilities/@sqw/symmetrise_sqw.m:127]: I believe it does not but I am not entirely sure. This is why it is left as it is -- symmetrisation work left unfinished.

File: horace_core/sqw/file_io/class_helpers/@pix_combine_info/pix_combine_info.m (hunk 3 above)
Comment (abuts) [reply horace_core/sqw/file_io/class_helpers/@pix_combine_info/pix_combine_info.m:256]: local range or local variable? The next row makes it entirely clear why it is called this way, as used for calculating `pix_range` (global one)

File: _test/test_mex_nomex/test_main_mex.m (hunk 7 above)
[the remark this answers is not in the document]
Comment (nickbattam-tessella) [reply _test/test_mex_nomex/test_main_mex.m]: the test should be split into two: one testing the matlab code, one testing the mex code. Then the 'skip' will correctly report the skipped section and not the whole test

[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_mex_nomex/test_main_mex.m]: the matlab code is tested against mex code, but some initial checks of Maltab code validity are performed too.

[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_mex_nomex/test_main_mex.m]: The test makes test-specific set up, runs Matlab code, checks the results sanity (low value but why not to check if its easy -- probably written when writing the test) and verifies Matlab results against mex results.  The main purpose is of course comparison of mex against nomex, so the purpose and location is valid. I do not believe there are any complexity in understanding the test and its results.

**Pull request merged**